dbms/src/Interpreters/JoinSwitcher.cpp

```diff
@@ -0,0 +1,92 @@
+#include <Common/typeid_cast.h>
+#include <Interpreters/JoinSwitcher.h>
+#include <Interpreters/Join.h>
+#include <Interpreters/MergeJoin.h>
+#include <Interpreters/join_common.h>
+
+namespace DB
+{
+
+static ColumnWithTypeAndName correctNullability(ColumnWithTypeAndName && column, bool nullable)
+{
+    if (nullable)
+        JoinCommon::convertColumnToNullable(column);
+    else
+        JoinCommon::removeColumnNullability(column);
+
+    return std::move(column);
+}
+
+JoinSwitcher::JoinSwitcher(std::shared_ptr<AnalyzedJoin> table_join_, const Block & right_sample_block_)
+    : switched(false)
+    , limits(table_join_->sizeLimits())
+    , table_join(table_join_)
+    , right_sample_block(right_sample_block_.cloneEmpty())
+{
+    join = std::make_shared<Join>(table_join, right_sample_block);
+
+    if (!limits.hasLimits())
+        limits.max_bytes = table_join->defaultMaxBytes();
+}
+
+bool JoinSwitcher::addJoinedBlock(const Block & block, bool)
+{
+    /// Trying to make MergeJoin without lock
+
+    if (switched)
```

Comment:
Should we name it 'is_merge_join'? Also TSan is going to complain, because the variable is plain bool and we read it without a lock. Should probably be atomic.

Reply:
Yup. I've made it atomic. The name 'switched' is better I think cause class is JoinSwitcher and join algo it switches to could be changed later.